### PR TITLE
Make docker build on FreeBSD

### PR DESCRIPTION
**Pull request by kvasdopil**
Disabling unsupported drivers, and make some packages compile on FreeBSD

### CODE REVIEW

**Comment by unclejack** [comment]
@kvasdopil There are a bunch of problems with your PR which need to be fixed in order to have this merged:
1) The PR can't be merged.
2) There are too many commits for these changes, please squash them down. A single commit should be enough.
3) The build tags for Linux only code shouldn't have `!freebsd` (aufs and all) as a build tag, they should have `linux` and `!linux` where required. example: `// +build !exclude_graphdriver_aufs,!freebsd` should be `// +build !exclude_graphdriver_aufs,linux` because aufs will only ever be on Linux. Some of these have already been changed for Windows.

Please update your PR to be based on the latest master.

**Comment by kvasdopil** [comment]
@unclejack thanks, already doing this.

What about removing those "import lxc" from daemon.go? it this ok?

**Comment by unclejack** [comment]
@kvasdopil Your timing with these changes couldn't have come at a better time because of the work to enable the building of the Docker daemon for Windows. The ongoing Windows work should have already taken care of that if I'm not mistaken and it should make things easier. If it's not taken care of on master, you should take a look at existing open PRs.

**Comment by kvasdopil** [comment]
@unclejack please take a look, seems like i finished

I think it's a good idea to remove extra build tags from _unix _windows _linux and _freebsd files, because go already filters them from being build on unsupported platforms, but i'd prefere to do that in separate branch.

**Comment by jfrazelle** [comment]
how are you compiling this? with hack/make.sh binary directly on the host?

**Comment by jfrazelle** [comment]
also can you give me an idea of what packages you have installed on the host other than go :) thanks

**Comment by kvasdopil** [comment]
@jfrazelle take a look at https://github.com/kvasdopil/docker/blob/freebsd-compat/FREEBSD-PORTING.md But remember that not all patches are in docker/master yet.

**Comment by kvasdopil** [comment]
@jfrazelle I mean we need last libnetwork and libcontainer for this pr to compile successfully

**Comment by jfrazelle** [comment]
ah awesome thanks :)

**Comment by kvasdopil** [comment]
So any progress on that?

**Comment by calavera** [comment]
Can you please rebase? we'll merge it when it turns :green_heart: again

**Comment by kvasdopil** [comment]
@calavera Done, but tests have failed for some reason. is it a race?

File: daemon/stats_freebsd.go [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+
+import (
+	"github.com/docker/docker/api/types"
+	"github.com/docker/libcontainer"
+)
+
+// convertStatsToAPITypes converts the libcontainer.Stats to the api specific
+// structs.  This is done to preserve API compatibility and versioning.
+func convertStatsToAPITypes(ls *libcontainer.Stats) *types.Stats {
+	// TODO Windows. Refactor accordingly to fill in stats.
```
Comment (cpuguy83) [inline daemon/stats_freebsd.go]: s/Windows/BSD

File: docker/daemon_unix.go [hunk 2]
```diff
@@ -11,7 +11,7 @@ import (
 	"github.com/docker/docker/daemon"
 	"github.com/docker/docker/pkg/system"
 
-	_ "github.com/docker/docker/daemon/execdriver/lxc"
+	//"github.com/docker/docker/daemon/execdriver/lxc"
```
Comment (cpuguy83) [inline docker/daemon_unix.go]: Shouldn't have commented code.
Does this import need to go into a file with linux build tag?

File: pkg/directory/directory_unix.go [hunk 3]
```diff
@@ -1,4 +1,4 @@
-// +build linux
+// +build linux freebsd
```
Comment (cpuguy83) [inline pkg/directory/directory_unix.go]: Can we remove these build tags now?

File: pkg/system/stat_freebsd.go [hunk 4]
```diff
@@ -0,0 +1,27 @@
+package system
```
Comment (cpuguy83) [inline pkg/system/stat_freebsd.go]: needs build tag?

File: docker/daemon_unix.go (hunk 2 above)
Comment (kvasdopil) [inline docker/daemon_unix.go]: I dont really get the point of that line, linux build seems to build ok without that.

File: pkg/directory/directory_unix.go (hunk 3 above)
Comment (kvasdopil) [inline pkg/directory/directory_unix.go]: I'm not sure if _unix suffix works as expected, it's not mentioned in go documentation, so maybe _unix files should have the tags?

File: pkg/system/stat_freebsd.go (hunk 4 above)
Comment (kvasdopil) [inline pkg/system/stat_freebsd.go]: _freebsd suffix works as expected, no need for build tag.

File: pkg/directory/directory_unix.go (hunk 3 above)
Comment (calavera) [inline pkg/directory/directory_unix.go]: yes, _unix files need tags. We have that convention to differentiate what works on "unix-ish" systems and what works on windows.

**Comment by calavera** [comment]
@kvasdopil there is an issue with running the integration tests that we're investigating, not really related with this PR.

**Comment by kvasdopil** [comment]
@calavera so how is it going?

**Comment by calavera** [comment]
@kvasdopil The issue in the integration tests should be fixed. But unfortunately we need to rebase this :pensive:

**Comment by thaJeztah** [comment]
@kvasdopil could you rebase this? Let's get the FreeBSD work moving again :+1:

**Comment by icecrime** [comment]
Ping @kvasdopil!

**Comment by kvasdopil** [comment]
@icecrime @thaJeztah hey guys i'm alive, thanks for ping. Had some serious problems with free time for last couple of weeks, and feel very sorry for that. I'll update my PRs asap.

**Comment by kvasdopil** [comment]
@icecrime @thaJeztah hello there. I'm back and i've updated the code (but having usual problems with ci ;( )

File: docker/daemon_unix.go (hunk 2 above)
Comment (tianon) [inline docker/daemon_unix.go]: It pulls in the driver code -- without it, lines like https://github.com/docker/docker/blob/06b922c63f660be431e4b678a50ae1ddb0dcf882/daemon/execdriver/lxc/init.go#L38 don't ever get executed, so it either needs to stay in here, or be moved to a Linux-specfic file in this same package.

Comment (tianon) [inline docker/daemon_unix.go]: This is similar to how [`database/sql`](https://golang.org/pkg/database/sql/) drivers usually work, too.

**Comment by calavera** [comment]
@kvasdopil it looks like this code needs some formatting. From the CI log:

```
---> Making bundle: validate-gofmt (in bundles/1.8.0-dev/validate-gofmt)
These files are not properly gofmt'd:
 - daemon/stats_freebsd.go
 - daemon/utils_freebsd.go

Please reformat the above files using "gofmt -s -w" and commit the result.
```

**Comment by kvasdopil** [comment]
@calavera looks like it's done

File: docker/daemon_unix.go (hunk 2 above)
Comment (cpuguy83) [inline docker/daemon_unix.go]: Still need to address this. Can probably just go in `daemon_linux.go` ?

File: daemon/stats_freebsd.go (hunk 1 above)
Comment (cpuguy83) [inline daemon/stats_freebsd.go]: Also need to update this comment to say freebsd, not windows.

**Comment by calavera** [comment]
There is an error building experimental, but I'm not sure it's completely related to these changes:

```
daemon/execdriver/native/template/default_template.go:6:2: no buildable Go source files in /go/src/github.com/docker/docker/vendor/src/github.com/opencontainers/runc/libcontainer/apparmor
Build step 'Execute shell' marked build as failure
```

**Comment by kvasdopil** [comment]
@calavera So what can i do about this? Other prs doesnt seem to have the same error, but i cant figure out how my modifications could trigger it.

**Comment by unclejack** [comment]
@kvasdopil `make cross` is failing on Linux with this PR.
```
Created symlinks: bundles/1.8.0-dev/cross/linux/amd64/docker bundles/1.8.0-dev/cross/linux/amd64/docker-1.8.0-dev bundles/1.8.0-dev/cross/linux/amd64/docker-1.8.0-dev.md5 bundles/1.8.0-dev/cross/linux/amd64/docker-1.8.0-dev.sha256
Building: bundles/1.8.0-dev/cross/linux/386/docker-1.8.0-dev
daemon/execdriver/native/template/default_template.go:6:2: no buildable Go source files in /go/src/github.com/docker/docker/vendor/src/github.com/opencontainers/runc/libcontainer/apparmor
make: *** [cross] Error 1
```

I'll post more comments if anything else fails for me. Tests are running on my local box.

File: docker/daemon_linux.go [hunk 5]
```diff
@@ -0,0 +1,5 @@
+package main
```
Comment (unclejack) [inline docker/daemon_linux.go]: @kvasdopil You need a build tag here: `// +build daemon`

**Comment by unclejack** [comment]
The tests didn't fail for me locally.

**Comment by calavera** [comment]
yes, flacky tests on windows are killing me.

**Comment by calavera** [comment]
okay, let's move this forward.

LGTM

**Comment by kvasdopil** [comment]
@calavera @unclejack @icecrime @thaJeztah so guys is there any chance to accept this pr?

**Comment by jfrazelle** [comment]
I say let's merge it if you all are comfortable, I'll set up the CI this week ;)

**Comment by thaJeztah** [comment]
Not really a Gopher, but would very much like to see this moving, so "lgtm" :-)

**Comment by tianon** [comment]
LGTM

**Comment by calavera** [comment]
This PR doesn't compile for me:

```
# github.com/docker/docker/pkg/reexec
pkg/reexec/command_freebsd.go:14: undefined: Self
```

You can copy what we do for windows:

```go
// Self returns the path to the current process's binary.
// Uses os.Args[0].
func Self() string {
	return naiveSelf()
}
```

**Comment by kvasdopil** [comment]
@calavera done

File: api/server/server_unix.go [hunk 6]
```diff
@@ -1,5 +1,3 @@
-// +build linux
```
Comment (jhowardmsft) [inline api/server/server_unix.go]: This needs either !windows or Linux FreeBSD.

File: builder/internals_unix.go [hunk 7]
```diff
@@ -1,5 +1,3 @@
-// +build linux
-
```
Comment (jhowardmsft) [inline builder/internals_unix.go]: !windows or Linux freebsd

File: daemon/daemon_no_aufs.go [hunk 8]
```diff
@@ -1,4 +1,4 @@
-// +build exclude_graphdriver_aufs,linux
+// +build exclude_graphdriver_aufs,linux !linux
```
Comment (jhowardmsft) [inline daemon/daemon_no_aufs.go]: This will pull in AUFS for Windows which is incorrect

File: daemon/graphdriver/driver_freebsd.go [hunk 9]
```diff
@@ -0,0 +1,8 @@
+package graphdriver
```
Comment (jhowardmsft) [inline daemon/graphdriver/driver_freebsd.go]: Does this need a //+build FreeBSD tag?

File: daemon/stats_freebsd.go [hunk 10]
```diff
@@ -0,0 +1,14 @@
+package daemon
```
Comment (jhowardmsft) [inline daemon/stats_freebsd.go]: Build tag?

File: daemon/utils_freebsd.go [hunk 11]
```diff
@@ -0,0 +1,9 @@
+package daemon
```
Comment (jhowardmsft) [inline daemon/utils_freebsd.go]: Build tag?

File: docker/daemon_unix.go [hunk 12]
```diff
@@ -11,7 +11,6 @@ import (
 	"github.com/docker/docker/daemon"
 	"github.com/docker/docker/pkg/system"
 
-	_ "github.com/docker/docker/daemon/execdriver/lxc"
```
Comment (jhowardmsft) [inline docker/daemon_unix.go]: It that really correct - removing the lxc driver on *nix? Not sure, but looks a little odd being in this PR.

File: pkg/parsers/operatingsystem/operatingsystem_freebsd.go [hunk 13]
```diff
@@ -0,0 +1,18 @@
+package operatingsystem
```
Comment (jhowardmsft) [inline pkg/parsers/operatingsystem/operatingsystem_freebsd.go]: Needs build tag

File: pkg/sysinfo/sysinfo_freebsd.go [hunk 14]
```diff
@@ -0,0 +1,7 @@
+package sysinfo
```
Comment (jhowardmsft) [inline pkg/sysinfo/sysinfo_freebsd.go]: Needs build tag

File: daemon/graphdriver/driver_freebsd.go (hunk 9 above)
Comment (kvasdopil) [inline daemon/graphdriver/driver_freebsd.go]: nope, _freebsd suffix works fine for freebsd

File: daemon/stats_freebsd.go (hunk 10 above)
Comment (kvasdopil) [inline daemon/stats_freebsd.go]: nope :)

File: daemon/utils_freebsd.go (hunk 11 above)
Comment (kvasdopil) [inline daemon/utils_freebsd.go]: no

File: pkg/parsers/operatingsystem/operatingsystem_freebsd.go (hunk 13 above)
Comment (kvasdopil) [inline pkg/parsers/operatingsystem/operatingsystem_freebsd.go]: still no

File: pkg/sysinfo/sysinfo_freebsd.go (hunk 14 above)
Comment (kvasdopil) [inline pkg/sysinfo/sysinfo_freebsd.go]: no tag needed :)

File: daemon/graphdriver/driver_freebsd.go (hunk 9 above)
Comment (jhowardmsft) [inline daemon/graphdriver/driver_freebsd.go]: Got it. Thanks for confirming. Ignore the comments then against _FreeBSD.go files.

File: docker/daemon_unix.go (hunk 12 above)
Comment (unclejack) [inline docker/daemon_unix.go]: It's just being moved to another separate source file.

Comment (jhowardmsft) [inline docker/daemon_unix.go]: Thanks for confirming.

File: daemon/daemon_no_aufs.go (hunk 8 above)
Comment (kvasdopil) [inline daemon/daemon_no_aufs.go]: didnt know aufs exist on windows, is it ok to rewrite it as // +build exclude_graphdriver_aufs freebsd ?

Comment (jhowardmsft) [inline daemon/daemon_no_aufs.go]: It doesn't exist on Windows :). I -think- you need to keep the original +build exclude_graphdriver_aufs,linux, but the second part depends on your intent - should this be included in FreeBSD?

Comment (unclejack) [inline daemon/daemon_no_aufs.go]: @jhowardmsft This adds a new file which makes this no_aufs source build on windows.

File: api/server/server_unix.go (hunk 6 above)
Comment (kvasdopil) [inline api/server/server_unix.go]: done

File: builder/internals_unix.go (hunk 7 above)
Comment (kvasdopil) [inline builder/internals_unix.go]: done

File: daemon/daemon_no_aufs.go (hunk 8 above)
Comment (jhowardmsft) [inline daemon/daemon_no_aufs.go]: daemon_no_aufs and daemon_aufs.go aren't currently being built on Windows with the tags as they are as the only call to migrateIfAufs is only made from daemon_unix.go which isn't compiled on Windows. Adding a !linux tag in daemon_no_aufs.go will mean daemon_no_aufs.go will be compiled on Windows which isn't necessary. I just verified that by deliberately putting compile errors in both daemon_no_aufs and daemon_aufs.go and successfully compiling the Windows daemon. I then added the tag above and sure enough, it tripped on the compile error I added to daemon_no_aufs. So I really don't believe this change is necessary here :)

**Comment by jhowardmsft** [comment]
LGTM from a Windows perspective now.

**Comment by calavera** [comment]
Tests pass, but CI is still very broken(unrelated to this). Code LGTM.

**Comment by kvasdopil** [comment]
So, what now?

**Comment by unclejack** [comment]
@jfrazelle You're assigned to this, can you please take a look and merge if you're OK with this PR?

**Comment by jfrazelle** [comment]
ugh i am so sorry it needs a rebase then i promise to merge once pinged

**Comment by kvasdopil** [comment]
@jfrazelle done :) ping!

**Comment by jfrazelle** [comment]
```

**Comment by jfrazelle** [comment]
got a go lint error

**Comment by kvasdopil** [comment]
@jfrazelle fixed (and i'm wondering why i haven't seen that earlier)

**Comment by jfrazelle** [comment]
test failures unrelated LGTM, merging and will get CI this week \o/

**Comment by thaJeztah** [comment]
Oh! Merged! Thanks @kvasdopil !!

**Comment by calavera** [comment]
:metal:

**Pull request merged**